src/hostapi/pulseaudio/pa_linux_pulseaudio_cb.c

```diff
@@ -0,0 +1,718 @@
+        && pa_stream_get_state( stream->outStream ) == PA_STREAM_READY )
+    {
+        pa_threaded_mainloop_lock(stream->mainloop);
+
+        /* Then we cancel all writing (if there is any)
+         *  and wait for disconnetion (TERMINATION) which
+         * can take a while for ethernet connections
+         */
+        pa_stream_cancel_write(stream->outStream);
+        pa_stream_disconnect(stream->outStream);
+
+        pa_threaded_mainloop_unlock(stream->mainloop);
+    }
+
+    if( stream->inStream != NULL
+        && pa_stream_get_state( stream->inStream ) == PA_STREAM_READY )
+    {
+        pa_threaded_mainloop_lock( stream->mainloop );
+
+        /* Then we disconnect stream and wait for
+         * Termination
+         */
+        pa_stream_disconnect( stream->inStream );
+
+        pa_threaded_mainloop_unlock( stream->mainloop );
+
+    }
+
+    /* Wait for termination for both */
+    while(!l_iLoop)
+    {
+        if( stream->inStream != NULL
+            && pa_stream_get_state( stream->inStream ) == PA_STREAM_TERMINATED )
+        {
+            pa_stream_unref( stream->inStream );
+            stream->inStream = NULL;
+
+            PaUtil_FreeMemory( stream->inBuffer );
+            stream->inBuffer = NULL;
+
+        }
+
+        if( stream->outStream != NULL
+            && pa_stream_get_state(stream->outStream) == PA_STREAM_TERMINATED )
+        {
+            pa_stream_unref( stream->outStream );
+            stream->outStream = NULL;
+
+            PaUtil_FreeMemory( stream->outBuffer );
+            stream->outBuffer = NULL;
+        }
+
+        if((stream->outStream == NULL
+           && stream->inStream == NULL)
+           || l_iError >= 5000 )
+        {
+              l_iLoop = 1;
+        }
+
+        l_iError ++;
+        usleep(100);
+    }
+
+    PaUtil_TerminateBufferProcessor( &stream->bufferProcessor );
+    PaUtil_TerminateStreamRepresentation( &stream->streamRepresentation );
+    PaUtil_FreeMemory(stream);
+
+    return result;
+}
+
+
+PaError PaPulseAudio_StartStreamCb( PaStream * s )
+{
+    PaError result = paNoError;
+    PaPulseAudio_Stream *stream = (PaPulseAudio_Stream *) s;
+    int streamStarted = 0;      /* So we can know whether we need to take the stream down */
+    PaPulseAudio_HostApiRepresentation *l_ptrPulseAudioHostApi = stream->hostapi;
+    const char *l_strName = NULL;
+    pa_operation *l_ptrOperation = NULL;
+
+    pa_threaded_mainloop_lock( l_ptrPulseAudioHostApi->mainloop );
+
+    /* Ready the processor */
+    PaUtil_ResetBufferProcessor( &stream->bufferProcessor );
+
+    /* Adjust latencies if that is wanted
+     * https://www.freedesktop.org/wiki/Software/PulseAudio/Documentation/Developer/Clients/LatencyControl/
+     *
+     * tlength is for Playback
+     * fragsize if for Record
+     */
+    stream->bufferAttr.maxlength = (uint32_t)-1;
+    stream->bufferAttr.tlength = pa_usec_to_bytes((pa_usec_t)stream->latency, &stream->outSampleSpec);
```

Comment:
This and line 507 cause assertion failures when dealing with streams that are only being opened in one direction. It's probably a good idea here to verify the validity of the streams involved and use inSampleSpec/outSampleSpec accordingly.

Reply:
I think I'll do that way that tlength is only used when playback as it's only for that and frag is only for records. After that I shouldn't cause problems. @tmiw you have some sample code in to make this happen?

Reply:
I think simply calling Pa_OpenStream() with outputParameters being NULL before Pa_StartStream() will do it based on the code we're currently using for the FreeDV project and the above line. inputParameters is assumed to be initialized properly, of course.